applications/vanilla/models/class.discussionmodel.php

```diff
@@ -1110,39 +1110,31 @@ public function calculate(&$discussion) {
 
         if (!property_exists($discussion, 'Read')) {
             $discussion->Read = !(bool)$discussion->CountUnreadComments;
-            if (!is_null($category['DateMarkedRead'])) {
+            if ($category && !is_null($category['DateMarkedRead'])) {
                 // If the category was marked explicitly read at some point, see if that applies here
-                if ($category['DateMarkedRead'] > $discussion->DateLastComment) {
-                    $discussion->Read = true;
-                }
-
-                if ($discussion->Read) {
-                    $discussion->CountUnreadComments = 0;
-                }
-            }
-
-            // Discussions are always unread to guests.
-            if (!Gdn::session()->isValid()) {
-                $discussion->Read = false;
+                $discussion->DateLastViewed = self::determineLastViewedDate($discussion->DateLastComment, $category['DateMarkedRead']);
             }
         }
 
-        // Logic for incomplete comment count.
-        if ($discussion->CountCommentWatch == 0 && $dateLastViewed = val('DateLastViewed', $discussion)) {
-            $discussion->CountUnreadComments = true;
-            if (Gdn_Format::toTimestamp($dateLastViewed) >= Gdn_Format::toTimestamp($discussion->LastDate)) {
-                $discussion->CountCommentWatch = $discussion->CountComments;
-                $discussion->CountUnreadComments = 0;
+        // Discussions are always unread to guests. Otherwise check Read status and Unread count.
+        if (!Gdn::session()->isValid()) {
+            $discussion->Read = false;
+        } else {
+            if (!is_null($discussion->DateLastViewed) && is_numeric($discussion->CountUnreadComments)) {
+                $updatedReadStatusAndUnreadCount = self::determineReadStatusAndUnreadCount(
+                    $discussion->CountComments,
+                    $discussion->DateLastComment,
+                    $discussion->CountCommentWatch,
+                    $discussion->DateLastViewed
```

Comment:
This is the max of `$category['DateMarkedRead']` and `$discussion->DateLastViewed`. Are you taking that into account? If so, can you clean out some of the `$category['DateMarkedRead']` stuff above?